Modernize functor_attachable and allow non-const default handlers

Neverlord opened the pull request:
I did some refactoring to use more C++17 constructs and also found an issue with the default handlers: passing a lambda declared as `mutable` to any of the `set_..._handler` functions fails to compile because the captured `fun` argument is `const` unless we also declare the lambdas that we use to construct the `std::function` object as `mutable`.

josephnoir reviewed (commented):
Should we add a test to check the claim of the change log? Just in case we temper with it again in the future.

Neverlord commented:
> Should we add a test to check the claim of the change log? Just in case we temper with it again in the future.

Yeah, you're right of course. Sorry for filing the PR without adding a test for it. Will do.

Neverlord commented:
@josephnoir the handlers are checked in several test suites. Duplicating them only for the sake of using a mutable lambda did not strike me as a good approach. So I've added a new test suite that only does compile-time checks to make sure the setters accept both mutable and const function objects. Also rebased while I'm at it.
